stratum/hal/lib/barefoot/bf_chassis_manager.cc

```diff
@@ -360,21 +404,50 @@ ::util::Status BFChassisManager::VerifyChassisConfig(
            << "Port groups are not supported on Tofino.";
   }
 
-  // If the class is initialized, we also need to check if the new config will
-  // require a change in the port layout. If so, report reboot required.
-  if (initialized_) {
-    std::map<uint64, std::map<uint32, PortKey>>
-        node_id_to_port_id_to_singleton_port_key;
+  // Validate singleton ports
+  std::map<uint64, std::map<uint32, PortKey>>
+      node_id_to_port_id_to_singleton_port_key;
+  {
+    std::map<uint64, int> node_id_to_unit;
+    // Map node ids to 0-based units
+    int unit(0);
+    for (auto& node : config.nodes()) {
+      node_id_to_unit[node.id()] = unit++;
+    }
 
+    ::util::Status status = ::util::OkStatus();  // errors to keep track of.
     for (const auto& singleton_port : config.singleton_ports()) {
       uint32 port_id = singleton_port.id();
       uint64 node_id = singleton_port.node();
 
+      PortKey singleton_port_key(singleton_port.slot(), singleton_port.port(),
+                                 singleton_port.channel());
       node_id_to_port_id_to_singleton_port_key[node_id][port_id] =
-          PortKey(singleton_port.slot(), singleton_port.port(),
-                  singleton_port.channel());
+          singleton_port_key;
+
+      // Make sure that the port exists by getting the SDK port ID
+      const int* unit = gtl::FindOrNull(node_id_to_unit, node_id);
+      if (unit == nullptr) {
+        APPEND_ERROR(status)
+            << "Node " << node_id << " not found for port " << port_id << ".";
+      } else {
+        uint32 unused_sdk_port_id;
+        APPEND_STATUS_IF_ERROR(
+            status, bf_pal_interface_->PortIdFromPortKeyGet(
+                        *unit, singleton_port_key, &unused_sdk_port_id));
+      }
+    }
+    if (!status.ok()) {
+      ::util::Status error = MAKE_ERROR(ERR_INVALID_PARAM)
+                             << "Error in singleton ports configuration.";
+      error.Update(status);
+      return error;
     }
```

Comment:
```suggestion
    RETURN_IF_ERROR(status)
            << "Error in singleton ports configuration.";
```

Reply:
I wanted an INVALID_PARAM error if there are any (rather than UNKNOWN), and couldn't figure out a better way to do that.

Reply:
`APPEND_ERROR` returns a new error object:

```
// Return a new error based on an existing error, with an additional string
// appended.  Otherwise behaves like MAKE_ERROR, including logging the error by
// default.
// Requires !status.ok().
// Example:
//   status = APPEND_ERROR(status) << ", more details";
//   return APPEND_ERROR(status) << ", more details";
#define APPEND_ERROR(status) \
  ::util::status_macros::MakeErrorStream((status), __FILE__, __LINE__)
```

Reply:
I believe the error code of the status doesn't change; it just appends to the error message. My goal was to make sure that the code of the returned error is ERR_INVALID_PARAM.

Also, I don't know if Update is doing what I want either.